File: lib/host/vimeo.c
/* 
* Copyright (C) 2009,2010 Toni Gundogdu.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "host.h"

/* Previously both flv and hd worked. Only hd seems to work now.
 * We'll default to hd as our new flv. */

#ifdef _1_
#define FLV_HD /* The old flv|hd scheme, disabled now by default. */
#endif

#ifdef FLV_HD
_host_constants(vimeo, "vimeo.com", "flv|hd");
#else
_host_constants(vimeo, "vimeo.com", "flv"); /* Make hd -> flv. */
#endif

_host_re(re_id,     "(?i)clip_id=(.*?)\"");
_host_re(re_title,  "(?i)<caption>(.*?)</caption>");
_host_re(re_sign,   "(?i)<request_signature>(.*?)</");
_host_re(re_exp,    "(?i)<request_signature_expires>(.*?)</");
#ifdef FLV_HD
_host_re(re_hd,     "(?i)<hd_button>(\\d+)</");
#endif

QUVIcode
handle_vimeo(const char *url, _quvi_video_t video) {
    char *content, *config_url, *config, *sign, *exp, *format;
    QUVIcode rc;
#ifdef FLV_HD
    int hd_avail;
#endif
    char *lnk;

#ifdef FLV_HD
    hd_avail = 0;
#endif

    /* host id */
    _host_id("vimeo");

    /* common */
    rc = parse_page_common(url, video, &content, re_id, NULL);

    _free(content);

    if (rc != QUVI_OK)
        return (rc);

    /* config */
    asprintf(&config_url,
        "http://vimeo.com/moogaloop/load/clip:%s", video->id);

    rc = fetch_to_mem(video, config_url, QUVISTATUSTYPE_CONFIG, &config);

    _free(config_url);

    if (rc != QUVI_OK)
        return (rc);

    /* title */
    _free(video->title);

    rc = regexp_capture(
        video->quvi,
        config,
        re_title,
        0,
        0,
        &video->title,
        0
    );

    if (rc != QUVI_OK) {
        _free(config);
        return (rc);
    }

    /* req. signature. */
    rc = regexp_capture(
        video->quvi,
        config,
        re_sign,
        0,
        0,
        &sign,
        0
    );

    if (rc == QUVI_OK) {
        /* req. signature expires */
        rc = regexp_capture(
            video->quvi,
            config,
            re_exp,
            0,
            0,
            &exp,
            0
        );

#ifdef FLV_HD
        if (rc == QUVI_OK) {
            QUVIcode _rc; /* do not override rc which we check further below */
            char *hd;

            _rc = regexp_capture(
                video->quvi,
                config,
                re_hd,
                0,
                0,
                &hd,
                0
            );

            if (hd && !strcmp(hd, "1"))
                hd_avail = 1;

            _free(hd);
        }
#endif
    }

    _free(config);

    if (rc != QUVI_OK) {
        _free(exp);
        _free(sign);
        return (rc);
    }

    /* video link */
    asprintf(&lnk,
        "http://vimeo.com/moogaloop/play/clip:%s/%s/%s"
#ifndef FLV_HD
        "/?q=hd&type=local&embed_location="
#endif
        ,video->id, sign, exp);

    _free(sign);
    _free(exp);

#ifdef FLV_HD
    format = video->quvi->format;

    if (format) {
        if ((!strcmp(format, "hd")
            || !strcmp(format, "best"))
            && hd_avail)
        {
            char *tmp = strdup(lnk);
            _free(lnk);
            asprintf(&lnk, "%s/?q=hd", tmp);
            _free(tmp);
        }
    }
#endif

    rc = add_video_link(&video->link, "%s", lnk);

    _free(lnk);

    return (QUVI_OK);
}

File: share/lua/website/vimeo.lua

-- quvi
-- Copyright (C) 2010,2011  Toni Gundogdu <legatvs@gmail.com>
--
-- This file is part of quvi <http://quvi.sourceforge.net/>.
--
-- This library is free software; you can redistribute it and/or
-- modify it under the terms of the GNU Lesser General Public
-- License as published by the Free Software Foundation; either
-- version 2.1 of the License, or (at your option) any later version.
--
-- This library is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- Lesser General Public License for more details.
--
-- You should have received a copy of the GNU Lesser General Public
-- License along with this library; if not, write to the Free Software
-- Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
-- 02110-1301  USA
--

-- w/ HD: <http://vimeo.com/1485507>
-- no HD: <http://vimeo.com/10772672>

-- Identify the script.
function ident(self)
    package.path = self.script_dir .. '/?.lua'
    local C      = require 'quvi/const'
    local r      = {}
    r.domain     = "vimeo.com"
    r.formats    = "default|best" -- default=sd, best=hd
    r.categories = C.proto_http
    local U      = require 'quvi/util'
    r.handles    = U.handles(self.page_url, {r.domain}, {"/%d+$"})
    return r
end

-- Parse video URL.
function parse(self)
    self.host_id  = "vimeo"
    self.page_url = vimeofy(self.page_url)

    local _,_,s = self.page_url:find('vimeo.com/(%d+)')
    self.id = s or error("no match: video id")

    local config_url = "http://vimeo.com/moogaloop/load/clip:" .. self.id
    local config = quvi.fetch(config_url, {fetch_type = 'config'})

    if config:find('<error>') then
        local _,_,s = config:find('<message>(.-)\n')
        error( (s == nil) and "no match: error message" or s )
    end

    local _,_,s = config:find("<caption>(.-)</")
    self.title  = s or error("no match: video title")

    local _,_,s = config:find("<request_signature>(.-)</")
    local sign  = s or error("no match: request signature")

    local _,_,s = config:find("<request_signature_expires>(.-)</")
    local exp   = s or error("no match: request signature expires")

    local _,_,s    = config:find('<isHD>(%d)')
    local hd_avail = (s and s == '1') and 1 or 0

    -- default=sd, best=hd
    local r_fmt = self.requested_format
    r_fmt = (r_fmt == 'best' and hd_avail == 1) and 'hd' or 'sd'

    self.url = {
        string.format("http://vimeo.com/moogaloop/play/clip:%s/%s/%s/?q=%s",
            self.id, sign, exp, r_fmt)
    }

    return self
end

-- vimeofy the URL.
function vimeofy(url)
    url = url:gsub("player.", "") -- player.vimeo.com
    return url
end

-- vim: set ts=4 sw=4 tw=72 expandtab:
